en: add doc for deploying heterogeneous cluster (#747)
pingcap/docs-tidb-operator: en/deploy-heterogeneous-tidb-cluster.md

---
title: Deploy a Heterogeneous TiDB Cluster
summary: Learn how to deploy a heterogeneous cluster for an existing TiDB cluster.
---

# Deploy a Heterogeneous TiDB Cluster

This document describes how to deploy a heterogeneous cluster for an existing TiDB cluster.

## Prerequisites

* You already have a TiDB cluster. If not, refer to [Deploy TiDB in General Kubernetes](deploy-on-general-kubernetes.md).

## Deploy a heterogeneous cluster

> A heterogeneous cluster is a differentiated instance node for an existing TiDB cluster. You can create a heterogeneous TiKV cluster with different configurations and labels to facilitate hotspot scheduling, or create a heterogeneous TiDB cluster for OLTP and OLAP workloads respectively.

Reviewer: ```suggestion
A heterogeneous cluster creates differentiated instances for an existing TiDB cluster. You can create a heterogeneous TiKV cluster with different configurations and labels to facilitate hotspot scheduling, or create a heterogeneous TiDB cluster for OLTP and OLAP workloads respectively.
```

### Deploy the cluster monitoring component

Save the following configuration as the `tidbmonitor.yaml` file. Replace `<heterogeneous_cluster_name>` with the desired name of your heterogeneous cluster, and replace `<origin_cluster_name>` with the name of the existing cluster.

{{< copyable "shell-regular" >}}

```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbMonitor
metadata:
  name: heterogeneous
spec:
  clusters:
    - name: ${origin_cluster_name}
    - name: ${heterogeneous_cluster_name}
  prometheus:
    baseImage: prom/prometheus
    version: v2.11.1
  grafana:
    baseImage: grafana/grafana
    version: 6.1.6
  initializer:
    baseImage: pingcap/tidb-monitor-initializer
    version: v4.0.6
  reloader:
    baseImage: pingcap/tidb-monitor-reloader
    version: v1.0.1
  imagePullPolicy: IfNotPresent
```

Execute the following command to create the heterogeneous cluster:

{{< copyable "shell-regular" >}}

```shell
kubectl create -f tidbmonitor.yaml -n ${namespace}
```

## Deploy a TLS-enabled heterogeneous cluster

> To enable TLS for a heterogeneous cluster, you need to explicitly declare TLS configuration, create a new `Secret` certificate, and issue the certificate using the same certification authority (CA) as the target cluster.

Reviewer: ```suggestion
To enable TLS for a heterogeneous cluster, you need to explicitly declare the TLS configuration, issue the certificates using the same certification authority (CA) as the target cluster and create new secrets with the certificates.
```

> If you want to issue the certificate using `cert-manager`, choose the same `Issuer` as the target cluster to create your `Certificate`.

Reviewer 2: ```suggestion
If you want to issue the certificate using `cert-manager`, choose the same `Issuer` as that of the target cluster to create your `Certificate`.
```

Reviewer: is it better to move L170 & 171 to this section and state that users can refer to those two procedures to create certs for the heterogeneous cluster?

Author: Sure, updated.

### Create a TLS-enabled heterogeneous cluster

> {{< copyable "shell-regular" >}}

Author: I'm a little confused about this YAML file. There's no clue about what users should do with it.

Reviewer: This is an example yaml that users can refer to.

Author: Can you explain in more details what users should do? Should they save the file, and execute some commands?

Reviewer: OK, the same procedure for the section `### Create a heterogeneous cluster`, save to file and `create -f`

Author: PTAL https://github.com/pingcap/docs-tidb-operator/pull/747/commits/c0a5d90d7a94a0fc49e5972f4ddd8a3a70c4525e

- [Enable TLS for the MySQL Client](enable-tls-for-mysql-client.md)

### Create a TLS-enabled heterogeneous cluster

{{< copyable "shell-regular" >}}

```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: heterogeneous
spec:
  tlsCluster:
    enabled: true
  configUpdateStrategy: RollingUpdate
  version: v4.0.6
  timezone: UTC
  pvReclaimPolicy: Delete
  discovery: {}
  cluster:

>     name: basic

Reviewer: ```suggestion
    name: ${origin_cluster_name}
```

    requests:
      storage: "1Gi"
    config:
      storage:
        # In basic examples, we set this to avoid using too much storage.
        reserve-space: "0MB"
  tidb:
    baseImage: pingcap/tidb
    replicas: 1
    service:
      type: ClusterIP
    config: {}
    tlsClient:
      enabled: true
  tiflash:
    baseImage: pingcap/tiflash
    maxFailoverCount: 1
    replicas: 1
    storageClaims:
      - resources:
          requests:
            storage: 1Gi
        storageClassName: standard
```

`spec.tlsCluster.enabled`: Determines whether to enable TLS between the components.
`spec.tidb.tlsClient.enabled`: Determines whether to enable TLS for MySQL client.

## See also

- [Enable TLS between TiDB Components](enable-tls-between-components.md)
- [Enable TLS for the MySQL Client](enable-tls-for-mysql-client.md)

> For the detailed configuration of a TLS-enabled heterogeneous cluster, see ['heterogeneous-tls'](https://github.com/pingcap/tidb-operator/tree/master/examples/heterogeneous-tls).

Reviewer: ```suggestion
For the detailed configuration of a TLS-enabled heterogeneous cluster, see ['heterogeneous-tls'](https://github.com/pingcap/tidb-operator/tree/master/examples/heterogeneous-tls) example.
```
